File: data/src/main/java/com/ahn/data/repository/FavoriteCurrencyRepositoryImpl.kt
package com.ahn.data.repository

import com.ahn.data.local.FavoriteCurrencyDataSource
import com.ahn.domain.repository.FavoriteCurrencyRepository
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject

class FavoriteCurrencyRepositoryImpl @Inject constructor(
    private val dataSource: FavoriteCurrencyDataSource
) : FavoriteCurrencyRepository {

    override fun getFavoriteCurrencyCodes(): Flow<List<String>> {
        return dataSource.getFavoriteCodes()
    }

    override suspend fun addFavorite(currencyCode: String) {
        dataSource.addFavorite(currencyCode)
    }

    override suspend fun removeFavorite(currencyCode: String) {
        dataSource.removeFavorite(currencyCode)
    }

    override suspend fun isFavorite(currencyCode: String): Boolean {
        return dataSource.isFavorite(currencyCode)
    }


}

File: ios/Packages/Data/Sources/Data/Favorite/Repository/FavoriteCurrencyRepositoryImpl.swift
import Domain

public struct FavoriteCurrencyRepositoryImpl: FavoriteCurrencyRepository {
    private let dataSource: FavoriteCurrencyDataSource

    public init(dataSource: FavoriteCurrencyDataSource) {
        self.dataSource = dataSource
    }

    public func favoriteCurrencyCodes() -> AsyncStream<[String]> {
        dataSource.favoriteCodes()
    }

    public func addFavorite(_ currencyCode: String) async throws {
        await dataSource.addFavorite(currencyCode)
    }

    public func removeFavorite(_ currencyCode: String) async throws {
        await dataSource.removeFavorite(currencyCode)
    }

    public func isFavorite(_ currencyCode: String) async throws -> Bool {
        await dataSource.isFavorite(currencyCode)
    }
}
